src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -4110,6 +4169,79 @@ impl Coordinator {
     }
 }
 
+#[derive(Debug)]
+struct CachedStatisticsOracle {
+    cache: BTreeMap<GlobalId, usize>,
+}
+
+const OPTIMIZER_MAX_STATS_WAIT: Duration = Duration::from_millis(250);
```

Comment:
We'll have to distinguish between one-off SELECTs and long-running dataflows, and have a much lower max wait for one-off SELECTs.

Ideally, we would also distinguish between fast-path one-off SELECTs and normal one-off SELECTs, but this is tricky, because we normally don't know that something is fast-path until the end of MIR optimization. Maybe we could have some heuristic already based on the HIR, e.g., if no joins then might be fast-path, so then don't wait more than 5ms or so.

Fast-path SELECTs are aiming for returning in something like 10ms, and normal one-off SELECTs can often return in something like 50-200ms.

Reply:
How's this plan?

 - tell statistics_oracle whether the query is one shot or not
   + 10ms for oneshots
   + 250ms for other queries
 - leave a TODO(mgree) to try to detect fast-path queries

Reply:
Sounds good! Could you please also mention this extra time in some kind of documentation of the switch that turns this on?